x-pack/plugins/maps/public/classes/layers/vector_layer/vector_layer.tsx

```diff
@@ -692,6 +697,69 @@ export class AbstractVectorLayer extends AbstractLayer implements IVectorLayer {
     return undefined;
   }
 
+  _createMasks() {
+    const masks: Mask[] = [];
+    const source = this.getSource();
+    if ('getMetricFields' in (source as IESAggSource)) {
+      const metricFields = (source as IESAggSource).getMetricFields();
+      metricFields.forEach((metricField) => {
+        const maskDescriptor = metricField.getMask();
+        if (maskDescriptor) {
+          masks.push(
+            new Mask({
+              esAggField: metricField,
+              isGeometrySourceMvt: source.isMvt(),
+              ...maskDescriptor,
+            })
+          );
+        }
+      });
+    }
+
+    this.getValidJoins().forEach((join) => {
+      const rightSource = join.getRightJoinSource();
+      if ('getMetricFields' in (rightSource as unknown as IESAggSource)) {
+        const metricFields = (rightSource as unknown as IESAggSource).getMetricFields();
+        metricFields.forEach((metricField) => {
+          const maskDescriptor = metricField.getMask();
+          if (maskDescriptor) {
+            masks.push(
+              new Mask({
+                esAggField: metricField,
+                isGeometrySourceMvt: source.isMvt(),
+                ...maskDescriptor,
+              })
+            );
+          }
+        });
+      }
+    });
+
+    return masks;
+  }
+
+  getMasks() {
+    return this._masks;
+  }
+
+  // feature-state is not supported in filter expressions
+  // https://github.com/mapbox/mapbox-gl-js/issues/8487
+  // therefore, masking must be accomplished via setting opacity paint property (hack)
+  _getAlphaExpression() {
+    const maskCaseExpressions: unknown[] = [];
+    this.getMasks().forEach((mask) => {
+      // case expressions require 2 parts
+      // 1) condition expression
+      maskCaseExpressions.push(mask.getMatchMaskedExpression());
+      // 2) output. 0 opacity styling "hides" feature
+      maskCaseExpressions.push(0);
+    });
+
+    return maskCaseExpressions.length
+      ? ['case', ...maskCaseExpressions, this.getAlpha()]
```

Comment:
Correct me if I'm wrong, but when there are multiple metrics with masks it looks like the mask is applied with a logical `AND` operator. I wonder if there will ever be a case where someone might want to apply the mask with an `OR` operator between multiple metrics?

Reply:
The [case expression](https://maplibre.org/maplibre-style-spec/expressions/#case) selects the first output whose corresponding test condition evaluates to true so its an `OR` relationship.

Reply:
Ahh, I see. Thanks for clarifying. 👍